accepted/2022/ReflectionInvoke.md

```diff
@@ -0,0 +1,474 @@
+- `object` (including boxing). Supports loose coupling scenarios are supported, like reflection today.
+- `ref <T>`. Supports new scenarios as mentioned earlier; type must be known ahead-of-time and due to no language support, cannot be a byref-like type like `Span<T>`.
+- `void*`. Unsafe cases used to support byref-like types in an unsafe manner.
+- `TypedReference`. Optional for now; pending language asks, it may make supporting byref-like types a safe operation.
+
+```cs
+namespace System.Reflection
+{
+     public ref struct MethodInvoker
+     {
+        // Zero-arg case:
+        public MethodInvoker()
+
+        // Variable-length number of arguments:
+        public unsafe MethodInvoker(ArgumentValue* argumentStorage, int argCount)
+
+        // Fixed length (say up to 8)
+        public MethodInvoker(ref ArgumentValuesFixed values)
+
+        // Dispose needs to be called with variable-length case
+        public void Dispose()
+
+        // Target
+        public object? GetTarget()
+        public ref T GetTarget<T>()
+
+        public void SetTarget(object value)
+        public void SetTarget(TypedReference value)
+        public unsafe void SetTarget(void* value, Type type)
+        public void SetTarget<T>(ref T value)
+
+        // Arguments
+        public object? GetArgument(int index)
+        public ref T GetArgument<T>(int index)
+
+        public void SetArgument(int index, object? value)
+        public void SetArgument(int index, TypedReference value)
+        public unsafe void SetArgument(int index, void* value, Type type)
+        public void SetArgument<T>(int index, ref T value)
+
+        // Return
+        public object? GetReturn()
+        public ref T GetReturn<T>()
+
+        public void SetReturn(object value)
+        public void SetReturn(TypedReference value)
+        public unsafe void SetReturn(void* value, Type type)
+        public void SetReturn<T>(ref T value)
+
+        // Invoke direct (limited validation and defaulting)
+        public unsafe void InvokeDirect(MethodBase method)
+
+        // Invoke (same validation and defaulting as reflection today)
+        public void Invoke(MethodBase method)
+     }
+
+     // This is used to define the correct storage requirements for the MethodInvoker variable-length cases.
+     // Internally it is based on 3 IntPtrs that are for 'ref', 'object value' and 'Type':
+     // - 'ref' points to either the 'value' location or a user-provided location with "void*", "ref <T>" or TypedReference.
+     // - 'object value' captures any user-provided object value in a GC-safe manner.
+     // - 'Type' is used in "void*", "ref <T>" and TypedReference cases for validation and in rare cases to prevent
+     //    Types from being GC'd.
+     public struct ArgumentValue { }
+```
+
+## ArgumentValuesFixed
+This class is used for cases where the known arguments are small.
+
+```cs
+namespace System.Reflection
+{
+    public ref partial struct ArgumentValuesFixed
```

Comment:
Why are you using this struct vs. a `Span<object>` which is backed by an `[InlineArray(8)]` struct that you slice to the appropriate length?

Reply:
That struct actually has 3 different types of storage (object, ByReference, and Type) and they need to be grouped something like this in order to register efficiently with the GC tracking support:
```cs
public unsafe ref struct ArgumentValuesFixed
  {
      internal object? _obj0;
      internal object? _obj1-to-lengthl

      internal RuntimeType? _type0;
      internal RuntimeType? _type1-to-lengthl

      internal ref byte _ref0;
      internal ref byte _ref1-to-lengthl
}
```

Also, I moved this type to the "future" section and will add it back, if necessary, based on perf.